proposals/distributed-async-epoch-reward.md

```diff
@@ -0,0 +1,237 @@
+---
+
+## Summary
+
+A new way to distribute epoch rewards asynchronously is proposed.
+
+## Motivation
+
+The distribution of epoch rewards at the beginning of an epoch becomes a
+significant bottleneck due to the rising number of stake accounts and voting
+nodes on the network.
+
+To address this bottleneck, we propose a new approach for distributing the
+epoch rewards asynchronously. Instead of distributing rewards at the start of
+the epoch, we propose an asynchronous distribution method that spreads the
+rewards payout over multiple blocks.
+
+
+## Detailed Design
+
+The major bottleneck is paying out rewards to stake accounts, which is about
+550K, while rewards for vote accounts are around 15K. Given the relatively
+small number of vote rewards, it make sense to keep vote rewards payout the
+same. They can still be paid out synchronously and efficiently at the first
+block of the epoch boundary. This reduce the impact on rewards for vote account
+and also simplify the over changes, and let us focus on solving the primary
+bottleneck - Stake Rewards. Stake rewards are going to be paid out
+asynchronously over multiple blocks.
+
+For async stake rewards payout, we will separate the computation of rewards
+from the actual distribution of rewards at the epoch boundary by dividing the
+process into two distinct phases: (1) rewards calculation, and (2) rewards
+distribution.
+
+
+### Rewards Calculation
+
+In contrast to the current method of computing epoch rewards, which involves
+directly computing rewards at the first slot of the epoch boundary in the bank
+runtime, the new approach involves moving the reward computation to a
+background service known as the `EpochRewardCalculationService`. This service
+starts automatically when a validator starts and listens to a designated
+channel for any incoming rewards calculation requests, performing the necessary
+rewards calculations accordingly.
+
+When a block first crosses the epoch boundary, its bank will send a request to
+the `EpochRewardCalculationService` to start the process of epoch reward
+computation. This marks the start of the reward computation phase.
+
+However, multiple rewards computation requests can be generated when there are
+forks at the epoch boundaries.
+
+The following diagram illustrates a fork happening at the epoch boundary `N`.
+
+```
+N-1 -- N -- N+1
+     \
+      \
+        N+2
+```
+
+Two rewards calculation requests are sent out at slot `N` and`N+2`, since they
+both cross the epoch boundary and are on different forks. But the second
+computation is not necessary because they share the same parent node before the
+start of the epoch. The inputs, vote_accounts/stake_account, to the rewards
+calculation will be identical.
+
+To avoid this repeated computation, the signature of the computation requests
+is calculated.
+
+```
+calculation_signature = hash(epoch_number, [&stake_accounts], [&vote_accounts], [&stake_delegations])
```

Comment:
yes.

Reply:
ok cool. do you have a sense for now long it'll take to compute this hash, especially with 600k+ stake accounts? 
as a nit, I'm not sure we need `epoch_number` in the `calculation_signature`. also wouldn't `stake_delegations` be essentially covered already by hashing the full contents of the stake and vote accounts

Reply:
Yes. epoch_number is not needed for the signature computation. 

The hash is only computed from the `[vote_account]` and `[stake_account]`. 

One implementation is to compute individual hash for [vote_account, delegated_stake accounts]. Then hash them together to get the final hash as the signature.

Maybe this can be simplified to collect all vote_accounts and stake_accounts into one array then hash the array at once, since all delegation info is captured in the stake_accounts? 

I will measure how long it takes to hash 600K stake accounts. If that takes too long, we could parallelize the hash computation.

```
    pub fn compute_rewards_calc_signature(
        &self,
        vote_with_stake_delegations_map: &DashMap<Pubkey, VoteWithStakeDelegations>,
    ) -> Hash {
        // compute individual vote/stake accounts hash
        let mut hashes: Vec<(Pubkey, Hash)> = Vec::new();

        vote_with_stake_delegations_map.iter().for_each(|x| {
            let mut hasher = Hasher::default();

            let vote_pubkey = x.key();
            let vote_delegation = x.value();
            let VoteWithStakeDelegations {
                vote_state,
                vote_account,
                delegations,
            } = vote_delegation;

            // compute vote account hash
            let vote_state_data = serialize(vote_state).unwrap();
            let vote_hash = crate::accounts_db::AccountsDb::hash_account_state(
                &vote_state_data[..],
                vote_account,
                vote_pubkey,
            );
            hasher. Update(vote_hash.as_ref());

            // compute delegation stake account hash (sorted by stake_pubkeys)
            for (stake_pubkey, stake_account) in delegations.iter().sorted_by_key(|x| x.0) {
                let stake_state_data = serialize(stake_account.stake_state()).unwrap();
                let hash = crate::accounts_db::AccountsDb::hash_account_state(
                    &stake_state_data[..],
                    stake_account.account(),
                    stake_pubkey,
                );
                hasher. Update(hash.as_ref());
            }

            let hash = crate::accounts_db::AccountsDb::to_hash(hasher);
            hashes. Push((*vote_pubkey, hash));
        });

        // compute final hash (sorted by vote_pubkeys)
        let mut hasher = Hasher::default();
        for (_key, hash) in hashes.iter().sorted_by_key(|x| x.0) {
            hasher. Update(hash.as_ref());
        }
        let result = crate::accounts_db::AccountsDb::to_hash(hasher);
        info!(
            "rewards_calc_request_signature: {}, slot {}",
            result,
            self. Slot()
        );
        result
    }
```

Reply:
It takes 15s to compute the hash for 600K stake accounts. :(

Reply:
RIP. What if we perform the background rewards calc for all new banks with a parent in the previous epoch. This may result in the identical computation for some forks across the boundary, but less of an issue since it's not synchronous.

We could reduce the duplicate computation by using a strategy of chunking the background rewards calc across blocks for the fork. So say we want the rewards credit to occur at block height 100 after the epoch boundary, each new block post epoch will cause 1% of the background rewards calc for that fork. If a fork is abandoned, then the background rewards calc for that fork is also abandoned

Reply:
Yeah. Complete Hash is too expensive. 

One idea is to use the slot of the last parent block before the epoch as key. All bocks forked from the same last parent in last epoch should produce the same rewards result (in the same way that `epoch_stakes` was populated as @brooksprumo pointed out) .